### PR TITLE
STS Storage Location Docs

### PR DESCRIPTION
**Pull request by DwayneJengSage**
https://sagebionetworks.jira.com/browse/PLFM-6012

User-facing documention for upcoming STS Storage Location (temporary AWS S3 credentials) feature.

Note: We discovered a bug in the feature that prevents it from working as intended. See https://sagebionetworks.jira.com/browse/PLFM-6008

### CODE REVIEW

File: articles/sts_storage_locations.md
Comment (brucehoff) [inline articles/sts_storage_locations.md]: instead of 'directly at' one might say 'Allow your compute cluster to read S3 objects using the S3 APIs

File: articles/sts_storage_locations.md
Comment (brucehoff) [inline articles/sts_storage_locations.md]: TRUE or T, not true

File: articles/sts_storage_locations.md
Comment (brucehoff) [inline articles/sts_storage_locations.md]: library(rjson)

**Comment by brucehoff** [comment]
Could you please add a section on how to create a Synapse File for an S3 object which was uploaded to an STS-storage location via the S3 API?  Part of the process is understanding how to properly choose a key for the S3 object.

**Comment by DwayneJengSage** [comment]
> Could you please add a section on how to create a Synapse File for an S3 object which was uploaded to an STS-storage location via the S3 API? Part of the process is understanding how to properly choose a key for the S3 object.

@brucehoff The section "Custom STS Storage Locations" link to https://docs.synapse.org/articles/custom_storage_location.html#adding-files-in-your-s3-bucket-to-synapse for how to add a file from S3 to Synapse.

There is also some sample code soon to be added to "Migrating Existing Data From an Existing S3 Bucket" which explains how to bulk migrate your data from S3.

Is there anything more that you think I need to add?

File: articles/sts_storage_locations.md
Comment (kdaily) [inline articles/sts_storage_locations.md]:
```suggestion
title: "AWS Security Token Service Storage Locations"
```

File: articles/custom_storage_location.md
Comment (kdaily) [inline articles/custom_storage_location.md]:
```suggestion
[AWS Security Token Service Storage Locations](sts_storage_locations.md)
```

File: articles/sts_storage_locations.md
Comment (kdaily) [reply articles/sts_storage_locations.md]: I would rename the file to `aws_sts_storage_locations.md` as well for linking purposes.

File: articles/sts_storage_locations.md
Comment (kdaily) [inline articles/sts_storage_locations.md]:
```suggestion
category: administration-and-settings
```

Comment (kdaily) [reply articles/sts_storage_locations.md]: We re-orged categories recently.

**Review by kdaily — CHANGES_REQUESTED** [review]
There's still some work to do with this that we can help with. The initial requested changes should be straightforward! I see that we need to write Python and Rexamples for obtaining temp credentials (using direct calls to the REST endpoints). I don't have the bandwidth right now to make those mods, so I'm going to leave this open. I would say we can get it done by next Monday (March 30) though.

**Comment by DwayneJengSage** [comment]
Made changes according to comments. Thanks!

Unfortunately, I don't know enough Python or R to make code examples for Python or R. The samples in the docs are modified from existing examples.

**Comment by DwayneJengSage** [comment]
@kdaily @kelshmo @Aryllen The corresponding feature has been pushed out to production, stack 305. In the web UI, it's hidden behind the alpha flag, but the API is available in the REST API and the Java client. (Python client TBD https://sagebionetworks.jira.com/browse/SYNPY-1049)

Is there anything else you need from me? Thanks!

**Comment by Aryllen** [comment]
Thanks for the information, @DwayneJengSage! I am hoping to have some time later this week or early next week to dive into this. The goal is for me to be able to add the R and Python examples before we merge this in. I'll definitely reach out if I have any questions.

**Comment by Aryllen** [comment]
This looks complete. I believe the only thing we need is for someone to verify the correctness of the R and Python examples that were adapted to work with this feature. @kelshmo is going to request assistance with this verification.

File: _articles/aws_sts_storage_locations.md
Comment (thomasyu888) [inline _articles/aws_sts_storage_locations.md:42]: What is the name of the bucket created?  You will need that to download files via s3

File: _articles/aws_sts_storage_locations.md
Comment (thomasyu888) [inline _articles/aws_sts_storage_locations.md:142]: Python equivalent here:

```

import synapseclients
import boto3

syn = synapseclient.login()
sts_credentials = syn.restGET(f"/entity/{FOLDER}/sts?permission=read_only")
client = boto3.client(
    's3',
    aws_access_key_id=sts_credentials['accessKeyId'],
    aws_secret_access_key=sts_credentials['secretAccessKey'],
    aws_session_token=sts_credentials['sessionToken'],
)
client.download_file('mybucket', 'hello.txt', '/tmp/hello.txt')
# mybucket -- The name of the bucket to download from.
# hello.txt -- The name of the key to download from.
# /tmp/hello.txt -- The path to the file to download to.

```

But I would like to provide a working example to actually download the file

File: _articles/aws_sts_storage_locations.md
Comment (philerooski) [reply _articles/aws_sts_storage_locations.md:42]: Is it returned in the POST call to `/storageLocation`? Looks like a BucketStorageLocationSetting has the bucket name but I'm not sure if that's the implementation returned in this case.

Comment (thomasyu888) [reply _articles/aws_sts_storage_locations.md:42]: It's not. There is a get `/storageLocation` but currently the s3 bucket name isn't exposed

Comment (thomasyu888) [reply _articles/aws_sts_storage_locations.md:42]: Edit:  There is a way to get it by just doing a `syn.get(..., downloadFile=False)`

File: _articles/aws_sts_storage_locations.md
Comment (jkiang13) [inline _articles/aws_sts_storage_locations.md:67]: I see how it is possible to add FileHandles for files uploaded via STS tokens (or otherwise) to an external bucket using the **POST /externalFileHandle/s3** endpoint. However for Synapse storage (i.e. non-external), after getting an STS token and using it to upload a file directly to a Synapse managed S3 bucket I don't see an equivalent way to be able to add a FileHandle representation for that uploaded file. The externalFileHandle/s3 endpoint doesn't work properly rejecting it as a non external storage location.

Also, per @thomasyu888 's point about the accessibility of the bucket name, to upload a file at all to Synapse storage using an STS token and an external S3 client one must already know the bucket name as it doesn't seem to be accessible via the upload destination or storage location.

File: _articles/aws_sts_storage_locations.md
Comment (DwayneJengSage) [reply _articles/aws_sts_storage_locations.md:42]: I believe (but need to confirm for sure) that you can only get the File Handle back if you're the File Handle owner. Which means it will work for your own File Handles, but not for File Handles that are shared with you.

I've filed https://sagebionetworks.jira.com/browse/PLFM-6226 to make this improvement.

File: _articles/aws_sts_storage_locations.md
Comment (DwayneJengSage) [reply _articles/aws_sts_storage_locations.md:67]: Good catch. I've filed https://sagebionetworks.jira.com/browse/PLFM-6227 to address this.

File: _articles/aws_sts_storage_locations.md
Comment (brucehoff) [reply _articles/aws_sts_storage_locations.md:42]: @DwayneJengSage Can you confirm that the destination['baseKey'] is the root of the scope of the bucket to which the STS token allows access?

File: _articles/aws_sts_storage_locations.md
Comment (brucehoff) [inline _articles/aws_sts_storage_locations.md:73]: "...will give users access to the whole bucket".  Is this really true or is it instead that the storage location has a root key prefix like <userid>/<uuid> that the STS token is constrained to?

File: _articles/aws_sts_storage_locations.md
Comment (DwayneJengSage) [reply _articles/aws_sts_storage_locations.md:42]: Yes, baseKey is the root folder to which the STS token allows access. It's an S3 prefix, so it doesn't contain the bucket name.

File: _articles/aws_sts_storage_locations.md
Comment (DwayneJengSage) [reply _articles/aws_sts_storage_locations.md:73]: This is true. If the baseKey is not specified, STS grants access to the whole bucket.

We can prevent this by auto-generating a baseKey, but this seems needlessly restrictive as users may *want* to grant access to the whole bucket.

File: _articles/aws_sts_storage_locations.md
Comment (kelshmo) [reply _articles/aws_sts_storage_locations.md:142]: @thomasyu888 can this code example live in https://github.com/Sage-Bionetworks/synapsePythonClient/pull/746/files#?

File: _articles/aws_sts_storage_locations.md
Comment (kelshmo) [inline _articles/aws_sts_storage_locations.md:73]:
```suggestion
{% include important.html content="If a baseKey is **not** specified, the temporary AWS credentials vended by STS will give users access to the whole bucket. To prevent access to the whole bucket, enter a folder path in your bucket that all files in the storage location should go into as the baseKey. " %}
```

Comment (kelshmo) [reply _articles/aws_sts_storage_locations.md:73]: I amended the language to be stronger here, since this sounds like it could be a very undesired outcome.

**Review by kelshmo — COMMENTED** [review]
suggested change to alert.

**Comment by kelshmo** [comment]
@ryanluce72 This document has a lot of pending issues blocking it from being released. The upcoming python client 2.1 release references this doc, as the release includes support for enabled S3 storage locations using AWS STS credentials. Can you weigh in on how to prioritize/address related issues?

**Comment by DwayneJengSage** [comment]
Minor fixes identified in this document review are being addressed in https://github.com/Sage-Bionetworks/Synapse-Repository-Services/pull/4067

**Comment by DwayneJengSage** [comment]
The following feature changes related to these doc changes will be in Stack-314 (which releases on Friday June 12:

* STS API now returns the bucket and base key along with the STS credentials
* STS Write Access will be disallowed for Synapse storage (but is still available for external storage)

**Comment by jkiang13** [comment]
Anything left to do to merge this? The Python 2.1 release notes link to this so it would be good to merge prior to that.

**Comment by DwayneJengSage** [comment]
@kdaily @kelshmo @Aryllen Are there any more action items left on this or are we ready to merge? Thanks!

**Review by kelshmo — CHANGES_REQUESTED** [review]
I have a number of suggestions.

- What permissions does a user need in order to be able to set STS enabled storage on a folder? 

- I am concerned that STS can only be enabled on empty folders. What does that mean for data that already exists in Synapse managed storage?

File: _articles/aws_sts_storage_locations.md
Comment (DwayneJengSage) [reply _articles/aws_sts_storage_locations.md:71]: "Custom" is what the external storage location page calls them, but internally, we call them external storage locations.

File: _articles/aws_sts_storage_locations.md
Comment (DwayneJengSage) [reply _articles/aws_sts_storage_locations.md:153]: This particular sample code is for populating an External Storage Location with an existing S3 bucket. I wrote it in Java because I don't know R or Python.

File: _articles/aws_sts_storage_locations.md
Comment (DwayneJengSage) [reply _articles/aws_sts_storage_locations.md:10]: Yes. I think bulk download is still one of the major scenarios here, and bulk download is supported for both Synapse storage and external S3 storage.

File: _articles/aws_sts_storage_locations.md
Comment (DwayneJengSage) [reply _articles/aws_sts_storage_locations.md]: I updated the comment with further explanation: "Synapse automatically generates folders for files uploaded through the Synapse website or clients. Files added directly to S3 may not match this folder structure. We recommend against mixing these two methods of adding files to prevent confusion in the folder structure."

**Comment by DwayneJengSage** [comment]
> What permissions does a user need in order to be able to set STS enabled storage on a folder?

This doesn't require any permissions other than creating storage locations and setting them on folders. Anyone can create storage locations. You need write access to a folder to set a storage location on it. I added some explanatory sentences.

File: _articles/aws_sts_storage_locations.md
Comment (kelshmo) [reply _articles/aws_sts_storage_locations.md]: The feature does not address this use case - https://sagebionetworks.jira.com/browse/PLFM-6282

**Review by kelshmo — APPROVED** [review]

**Pull request merged**
